File: PirateLexer/TokenRepository.cs
using System.Globalization;
using PirateLexer.Tokens;
using PirateLexer.Enums;
using PirateLexer.Interfaces;
using PirateLexer.Tokens.Interfaces;

namespace PirateLexer;

/// <summary>
/// A class which handles the complex logic of the lexer.
/// </summary>
public class TokenRepository : ITokenRepository
{
    private readonly IKeyWordService _KeyWordService;

    public TokenRepository(IKeyWordService KeyWordService)
    {
        _KeyWordService = KeyWordService;
    }

    public TokenResult MakeNumber(string text, int position)
    {
        var dotCount = 0;
        var numberString = string.Empty;
        Token token;

        while (Char.IsDigit(text[position]) || text[position] == '.')
        {
            if (text[position] == '.')
            {
                if (dotCount == 1)
                {
                    break;
                }
                dotCount += 1;
                numberString += '.';
            }
            else
            {
                numberString += text[position];
            }
            position += 1;
            if (position == text.Length) break;
        }

        if (dotCount == 0)
        {
            token = new Token(TokenGroup.VALUE, TokenType.INT, int.Parse(numberString));

        }
        else
        {
            CultureInfo cultureInfo = (CultureInfo)CultureInfo.CurrentCulture.Clone();
            cultureInfo.NumberFormat.CurrencyDecimalSeparator = "."; cultureInfo.NumberFormat.NumberDecimalSeparator = ".";
            token = new Token(TokenGroup.VALUE, TokenType.FLOAT, float.Parse(numberString, cultureInfo));
        }

        return new TokenResult()
        {
            Token = token,
            Position = position
        };
    }

    public TokenResult MakeIdentifier(string text, int position)
    {
        var idString = string.Empty;

        while (Char.IsLetter(text[position]) || !Char.IsNumber(text[position]) || !Char.IsWhiteSpace(text[position]) || !Char.IsSeparator(text[position]))
        {
            idString += text[position];
            position += 1;
            if (position == text.Length) break;
            if (text[position] == '.')
            {
                idString += text[position];
                position += 1;
            }
            if (Char.IsNumber(text[position]) || Char.IsPunctuation(text[position]) || Char.IsWhiteSpace(text[position]) || Char.IsSeparator(text[position])) break;
        }

        var TokenTypeType = _KeyWordService.GetTypeKeyword(idString);
        if (TokenTypeType != TokenType.Empty)
        {
            return new TokenResult()
            {
                Token = new Token(TokenGroup.TYPEKEYWORD, TokenTypeType, idString),
                Position = position
            };
        }

        TokenTypeType = _KeyWordService.GetTokenControlKeyword(idString);

        if (TokenTypeType != TokenType.Empty)
        {
            return new TokenResult()
            {
                Token = new Token(TokenGroup.CONTROLKEYWORD, TokenTypeType, idString),
                Position = position
            };
        }

        return new TokenResult()
        {
            Token = new Token(TokenGroup.SYNTAX, TokenType.IDENTIFIER, idString),
            Position = position
        };
    }

    public TokenResult MakeString(string text, int position)
    {
        var resultString = string.Empty;
        var escapeCharacter = false;
        position += 1;

        Dictionary<string, string> escapeCharacters = new Dictionary<string, string>() { };
        escapeCharacters.Add("n", "\n");
        escapeCharacters.Add("t", "\t");

        while (text[position] != '"' || escapeCharacter)
        {
            if (escapeCharacter)
            {
                resultString += escapeCharacters[text[position].ToString()];
            }
            else
            {
                if (text[position] == '\\')
                {
                    escapeCharacter = true;
                }
                else
                {
                    resultString += text[position];
                }
            }
            position += 1;
            escapeCharacter = false;
            if (position == text.Length) break;
        }
        position += 1;

        return new TokenResult()
        {
            Token = new Token(TokenGroup.VALUE, TokenType.STRING, resultString),
            Position = position
        };
    }

    public TokenResult MakeChar(string text, int position)
    {
        position += 1;
        var resultString = text[position];
        position += 1;
        if (text[position] != '\'')
        {
            throw new InvalidOperationException("Char is not one letter");
        }
        position += 1;

        return new TokenResult()
        {
            Token = new Token(TokenGroup.VALUE, TokenType.CHAR, resultString),
            Position = position
        };
    }

    public TokenResult MakeNotEquals(string text, int position)
    {
        position += 1;

        if (text[position] == '=')
        {
            position += 1;
            return new TokenResult()
            {
                Token = new Token(TokenGroup.COMPARISONOPERATORS, TokenType.NOTEQUALS),
                Position = position
            };
        }

        position += 1;
        throw new Exception("Expected Character Error: '=' (after '!')");
    }

    public TokenResult MakeGreaterThan(string text, int position)
    {
        var tokenType = TokenType.GREATERTHAN;
        position += 1;

        if (text[position] == '=')
        {
            position += 1;
            tokenType = TokenType.GREATERTHANEQUALS;
        }

        return new TokenResult()
        {
            Token = new Token(TokenGroup.COMPARISONOPERATORS, tokenType),
            Position = position
        };

    }

    public TokenResult MakeLessThan(string text, int position)
    {
        var tokenType = TokenType.LESSTHAN;
        position += 1;

        if (text[position] == '=')
        {
            position += 1;
            tokenType = TokenType.LESSTHANEQUALS;
        }

        return new TokenResult()
        {
            Token = new Token(TokenGroup.COMPARISONOPERATORS, tokenType),
            Position = position
        };
    }

    public TokenResult MakeEquals(string text, int position)
    {
        Token token;
        position += 1;
        if (text[position] == '=')
        {
            position += 1;
            token = new Token(TokenGroup.COMPARISONOPERATORS, TokenType.DOUBLEEQUALS);
        }
        else
        {
            token = new Token(TokenGroup.SYNTAX, TokenType.EQUALS);
        }

        return new TokenResult()
        {
            Token = token,
            Position = position
        };
    }

    public TokenResult MakePlus(string text, int position)
    {
        Token token;
        position += 1;
        if (text[position] == '=')
        {
            position += 1;
            token = new Token(TokenGroup.SYNTAX, TokenType.PLUSEQUALS);
        }
        else
        {
            token = new Token(TokenGroup.OPERATORS, TokenType.PLUS);
        }

        return new TokenResult()
        {
            Token = token,
            Position = position
        };
    }

    public TokenResult MakeDivide(string text, int position)
    {
        Token token;
        position += 1;
        if (text[position] == '/')
        {
            position += 1;
            token = new Token(TokenGroup.SYNTAX, TokenType.DOUBLEDIVIDE);
        }
        else
        {
            token = new Token(TokenGroup.OPERATORS, TokenType.DIVIDE);
        }

        return new TokenResult()
        {
            Token = token,
            Position = position
        };
    }
}

File: Pirate.Lexer.F/TokenRepository.fs
namespace Pirate.Lexer

open Pirate.Lexer.Tokens
open Pirate.Lexer.Enums
open System
open System.Globalization
open System.Collections.Generic

exception TokenRepositoryException of string

type TokenRepository(keyWordService:KeyWordService) =
    let mutable _keyWordService = keyWordService

    member _.MakeNumber(text:string, position:int) : TokenResult =
        let mutable position = position

        let mutable dotCount = 0
        let mutable numberString = "";
        let mutable token = Token(VALUE, TokenType.Empty, 0)

        let mutable Break = false
        while not Break && (Char.IsDigit(text.[position]) || text.[position] = '.') do
            if text.[position] = '.' then
                if dotCount = 1 then
                    Break <- true
                else
                    dotCount <- dotCount + 1
                    numberString <- numberString + "."
            else
                numberString <- numberString + text.[position].ToString()

            position <- position + 1
            if position = text.Length then
                Break <- true

        if dotCount = 0 then
            token <- Token(VALUE, INT, int numberString)
        else
            let cultureInfo = CultureInfo.CurrentCulture.Clone() :?> CultureInfo
            cultureInfo.NumberFormat.CurrencyDecimalSeparator <- "."
            cultureInfo.NumberFormat.NumberDecimalSeparator <- "."
            
            match Double.TryParse(numberString, NumberStyles.Number, CultureInfo.InvariantCulture) with
            | (true, result) -> token <- Token(VALUE, FLOAT, result)
            | (false, result) -> raise (System.Exception("Invalid number format"))

        TokenResult(token, position)

    member _.MakeIdentifier(text:string, position:int) : TokenResult =
        let mutable position = position

        let mutable idString = "";

        let mutable Break = false
        try
            while not Break && (Char.IsLetter(text.[position]) && Char.IsDigit(text.[position]) = false && Char.IsWhiteSpace(text.[position]) = false && Char.IsSeparator(text.[position]) = false ) do
                idString <- idString + text.[position].ToString()
                position <- position + 1
                if text.[position] = '.' then
                    idString <- idString + text.[position].ToString()
                    position <- position + 1

                if position = text.Length then
                    Break <- true
                else if Char.IsNumber(text.[position]) || Char.IsDigit(text.[position]) || Char.IsWhiteSpace(text.[position]) || Char.IsSeparator(text.[position]) then
                    Break <- true
        with
        | :? IndexOutOfRangeException -> ()
        | _ -> raise (TokenRepositoryException("Invalid identifier format"))

        let mutable TokenTypeType = _keyWordService.GetTypeKeyword(idString)
        let mutable TokenTypeControl = _keyWordService.GetTokenControlKeyword(idString)
        if TokenTypeType <> TokenType.Empty then
            TokenResult(
                Token(TYPEKEYWORD, TokenTypeType, idString), 
                position
            )
        else if TokenTypeControl <> TokenType.Empty then
            TokenResult(
                Token(CONTROLKEYWORD, TokenTypeControl, idString), 
                position
            )
        else 
            TokenResult(
                Token(SYNTAX, IDENTIFIER, idString), 
                position
            )

    member _.MakeString(text:string, position:int) : TokenResult =
        let mutable position = position + 1

        let mutable resultString = "";
        let mutable escapeCharacter = false

        let escapeCharacters = new Dictionary<string, string>();
        escapeCharacters.Add("n", "\n");
        escapeCharacters.Add("t", "\t");

        let mutable Break = false
        while not Break && text.[position] <> '"' do
            if escapeCharacter then
                resultString <- resultString + escapeCharacters.[text.[position].ToString()]
            else
                if text.[position] = '\\' then
                    escapeCharacter <- true
                else
                    resultString <- resultString + text.[position].ToString()
            position <- position + 1
            escapeCharacter <- false
            if position = text.Length then
                Break <- true

        position <- position + 1

        TokenResult(
            Token(VALUE, STRING, resultString), 
            position
        )

    member _.MakeChar(text:string, position:int) : TokenResult =
        let mutable position = position + 1

        let mutable resultString = text.[position].ToString()
        position <- position + 1
        if text.[position] <> '\'' then
            raise (System.Exception("Invalid char format"))
        position <- position + 1

        TokenResult(
            Token(VALUE, CHAR, resultString), 
            position
        )

    member _.MakeNotEquals(text:string, position:int) : TokenResult =
        let mutable position = position + 1

        if text.[position] <> '=' then
            raise (System.Exception("Invalid not equals format"))

        position <- position + 1
        TokenResult(
            Token(COMPARISONOPERATORS, NOTEQUALS, "!="),
            position
        )

    member _.MakeGreaterThan(text:string, position:int) : TokenResult =
        let mutable position = position + 1

        try
            if text.[position] = '=' then
                position <- position + 1
                TokenResult(
                    Token(COMPARISONOPERATORS, GREATERTHANEQUALS, ">="),
                    position
                )
            else
                TokenResult(
                    Token(COMPARISONOPERATORS, GREATERTHAN, ">"),
                    position
                )
        with
        | :? System.IndexOutOfRangeException ->
            TokenResult(
                Token(COMPARISONOPERATORS, GREATERTHAN, ">"),
                position
            )
        | _ -> raise (System.Exception("Invalid greater than format"))

    member _.MakeLessThan(text:string, position:int) : TokenResult =
        let mutable position = position + 1
        try
            if text.[position] = '=' then
                position <- position + 1
                TokenResult(
                    Token(COMPARISONOPERATORS, LESSTHANEQUALS, "<="),
                    position
                )
            else
                TokenResult(
                    Token(COMPARISONOPERATORS, LESSTHAN, "<"),
                    position
                )
        with
        | :? System.IndexOutOfRangeException ->
            TokenResult(
                Token(COMPARISONOPERATORS, LESSTHAN, "<"),
                position
            )
        | _ -> raise (System.Exception("Invalid less than format"))

    member _.MakeEquals(text:string, position:int) =
        let mutable position = position + 1
        try
            if text.[position] = '=' then
                position <- position + 1
                TokenResult(
                    Token(COMPARISONOPERATORS, EQUALS, "=="),
                    position
                )
            else
                TokenResult(
                    Token(SYNTAX, EQUALS, "="),
                    position
                )
        with
        | :? System.IndexOutOfRangeException ->
            TokenResult(
                Token(SYNTAX, EQUALS, "="),
                position
            )
        | _ -> raise (System.Exception("Invalid equals format"))

    member _.MakePlus(text:string, position:int) =
        let mutable position = position + 1
        try
            if text.[position] = '=' then
                position <- position + 1
                TokenResult(
                    Token(SYNTAX, PLUSEQUALS, "+="),
                    position
                )
            else
                TokenResult(
                    Token(OPERATORS, PLUS, "+"),
                    position
                )
        with
        | :? System.IndexOutOfRangeException ->
            TokenResult(
                Token(OPERATORS, PLUS, "+"),
                position
            )
        | _ -> raise (System.Exception("Invalid plus format"))

    member _.MakeDivide(text:string, position:int) =
        let mutable position = position + 1

        try 
            if text.[position] = '/' then
                position <- position + 1
                TokenResult(
                    Token(SYNTAX, DOUBLEDIVIDE, "//"),
                    position
                )
            else
                TokenResult(
                    Token(OPERATORS, DIVIDE, "/"),
                    position
                )
        with
            | :? System.IndexOutOfRangeException ->
                TokenResult(
                    Token(OPERATORS, DIVIDE, "/"),
                    position
                )
            | _ -> raise (System.Exception("Invalid divide format"))